File: exercise/C#/day23/ControlSystem.Core/ControlSystem.cs
using System.Collections;
using ControlSystem.External;

namespace ControlSystem.Core;

public class System
{
    private const int XmasSpirit = 40;
    private readonly Dashboard _dashboard;
    private readonly MagicStable _magicStable;
    private readonly List<ReindeerPowerUnit> _reindeerPowerUnits;
    public SleighEngineStatus Status { get; set; }
    public SleighAction Action { get; set; }
    private float _controlMagicPower = 0;

    private readonly Amplifiers _amplifiers = Amplifiers.Build();
    
    public System(MagicStable magicStable)
    {
        _magicStable = magicStable;
        _dashboard = new Dashboard();
        _reindeerPowerUnits = BringAllReindeers();
    }

    private List<ReindeerPowerUnit> BringAllReindeers() =>
        _magicStable
            .GetAllReindeers()
            .OrderByDescending(r => r.GetMagicPower())
            .Select(reindeer => new ReindeerPowerUnit(reindeer,_amplifiers.GetNext()))
            .ToList();

    public void StartSystem()
    {
        _dashboard.DisplayStatus("Starting the sleigh...");
        Status = SleighEngineStatus.On;
        _dashboard.DisplayStatus("System ready.");
    }

    public void Ascend()
    {
        if (Status != SleighEngineStatus.On)
        {
            throw new SleighNotStartedException();
        }
        
        foreach (var reindeerPowerUnit in _reindeerPowerUnits)
        {
            _controlMagicPower += reindeerPowerUnit.HarnessMagicPower();
        }

        if (CheckReindeerStatus())
        {
            _dashboard.DisplayStatus("Ascending...");
            Action = SleighAction.Flying;
            _controlMagicPower = 0;
        }
        else throw new ReindeersNeedRestException();
    }

    public void Descend()
    {
        if (Status == SleighEngineStatus.On)
        {
            _dashboard.DisplayStatus("Descending...");
            Action = SleighAction.Hovering;
        }
        else throw new SleighNotStartedException();
    }

    public void Park()
    {
        if (Status != SleighEngineStatus.On) throw new SleighNotStartedException();
        
        _dashboard.DisplayStatus("Parking...");

        foreach (var reindeerPowerUnit in _reindeerPowerUnits)
        {
            reindeerPowerUnit.Reindeer.TimesHarnessing = 0;
        }

        Action = SleighAction.Parked;
    }

    public void StopSystem()
    {
        _dashboard.DisplayStatus("Stopping the sleigh...");
        Status = SleighEngineStatus.Off;
        _dashboard.DisplayStatus("System shutdown.");
    }

    private bool CheckReindeerStatus()
    {
        return _controlMagicPower >= XmasSpirit;
    }
}

File: exercise/ts/day23/src/control.core/controlSystem.ts
import {Dashboard} from "./dashboard";
import {MagicStable} from "../external/stable/magicStable";
import {ReindeerPowerUnit} from "./reindeerPowerUnit";
import {SleighAction, SleighEngineStatus} from "./enums";
import {Reindeer} from "../external/deer/reindeer";
import {ReindeersNeedRestException, SleighNotStartedException} from "./exceptions";

export class ControlSystem {
    private XmasSpirit = 40;
    private dashboard: Dashboard;
    private magicStable = new MagicStable();
    private reindeerPowerUnits: ReindeerPowerUnit[];
    public status: SleighEngineStatus;
    public action: SleighAction;
    private controlMagicPower = 0;

    constructor() {
        this.dashboard = new Dashboard();
        this.reindeerPowerUnits = this.bringAllReindeers();
    }

    private bringAllReindeers(): ReindeerPowerUnit[] {
        return this.magicStable.getAllReindeers().map(reindeer => this.attachPowerUnit(reindeer));
    }

    public attachPowerUnit(reindeer: Reindeer): ReindeerPowerUnit {
        return new ReindeerPowerUnit(reindeer);
    }

    public startSystem() {
        this.dashboard.displayStatus("Starting the sleigh...");
        this.status = SleighEngineStatus.ON;
        this.dashboard.displayStatus("System ready.");
    }

    public ascend() {
        if (this.status === SleighEngineStatus.ON) {
            this.reindeerPowerUnits.forEach(unit => {
                this.controlMagicPower += unit.harnessMagicPower();
            });

            if (this.checkReindeerStatus()) {
                this.dashboard.displayStatus("Ascending...");
                this.action = SleighAction.FLYING;
                this.controlMagicPower = 0;
            } else {
                throw new ReindeersNeedRestException();
            }
        } else {
            throw new SleighNotStartedException();
        }
    }

    public descend() {
        if (this.status === SleighEngineStatus.ON) {
            this.dashboard.displayStatus("Descending...");
            this.action = SleighAction.HOVERING;
        } else {
            throw new SleighNotStartedException();
        }
    }

    public park() {
        if (this.status === SleighEngineStatus.ON) {
            this.dashboard.displayStatus("Parking...");

            this.reindeerPowerUnits.forEach(unit => {
                unit.reindeer.timesHarnessing = 0;
            });

            this.action = SleighAction.PARKED;
        } else {
            throw new SleighNotStartedException();
        }
    }

    public stopSystem() {
        this.dashboard.displayStatus("Stopping the sleigh...");
        this.status = SleighEngineStatus.OFF;
        this.dashboard.displayStatus("System shutdown.");
    }

    private checkReindeerStatus(): boolean {
        return this.controlMagicPower >= this.XmasSpirit;
    }
}